nemo/collections/asr/parts/mixins/interctc_mixin.py

```diff
@@ -0,0 +1,140 @@
+    property and registers layer_output_X and layer_length_X for all layers that
+    we want to get loss from. Then call
+
+        * ``setup_interctc`` in the init method
+        * ``self.add_interctc_losses`` after computing regular loss.
+        * ``self.finalize_interctc_metrics(metrics, outputs, prefix="val_")``
+          in the `multi_validation_epoch_end` method.
+        * ``self.finalize_interctc_metrics(metrics, outputs, prefix="test_")``
+          in the `multi_test_epoch_end` method.
+    """
+
+    def setup_interctc(self):
+        self.intermediate_loss_weights = self.cfg.get('intermediate_loss_weights', [])
+        self.main_loss_weight = 1.0 - sum(self.intermediate_loss_weights)
+        if self.main_loss_weight <= 0.0:
+            raise ValueError(
+                "Make sure that sum of intermediate loss weights is < 1.0. "
+                "Note that we don't do any normalization and assign "
+                "remaining weight to the regular model loss. "
+                "E.g., if intermediate_loss_weights = [0.1, 0.3], regular "
+                "loss will have weight of 0.6"
+            )
+        self.interctc_enabled = len(self.intermediate_loss_weights) > 0
+        if self.interctc_enabled and not hasattr(self.encoder, 'capture_output_at_layers'):
+            raise ValueError('To use intermediate CTC loss, encoder has to define "capture_output_at_layers" property')
+
+        if len(self.encoder.capture_output_at_layers) != len(self.intermediate_loss_weights):
+            raise ValueError('Length of encoder.capture_output_at_layers has to match intermediate_loss_weights')
+
+    def finalize_interctc_metrics(self, metrics, outputs, prefix):
+        if self.interctc_enabled:
+            for layer_idx in self.encoder.capture_output_at_layers:
+                loss = torch.stack([x[f"{prefix}inter_ctc_loss_l{layer_idx}"] for x in outputs]).mean()
+                wer_num = torch.stack([x[f"{prefix}inter_wer_num_l{layer_idx}"] for x in outputs]).sum()
+                wer_denom = torch.stack([x[f"{prefix}inter_wer_denom_l{layer_idx}"] for x in outputs]).sum()
+                metrics["log"].update(
+                    {
+                        f"{prefix}inter_ctc_loss_l{layer_idx}": loss,
+                        f"{prefix}inter_wer_l{layer_idx}": wer_num / wer_denom,
+                    }
+                )
+            metrics["log"][f"{prefix}final_ctc_loss"] = torch.stack(
+                [x[f"{prefix}final_ctc_loss"] for x in outputs]
+            ).mean()
+
+    def get_captured_tensors(self):
+        if not self.interctc_enabled:
+            return []
+        # if intermediate_loss_weights was set, the encoder has to register
+        # layer_output_X and layer_length_X tensors. We need to apply decoder
+        # to each of them and compute CTC loss.
+        module_registry = AccessMixin.get_module_registry(self.encoder)['']  # key for encoder
```

Comment:
[''] ?

Reply:
That's the key that AccessMixin has when assigning tensors to the current module. I don't have control over it :)